atc/builds/planner_test.go

```diff
@@ -0,0 +1,702 @@
+package builds_test
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/concourse/concourse/atc"
+	"github.com/concourse/concourse/atc/builds"
+	"github.com/concourse/concourse/atc/db"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
```

Comment:
Can we stick with the same test suite within a given package?

Reply:
Yeah it's a little hokey that it doesn't output the planner tests at all when they pass. I think I'll just convert the tracker tests.

Reply:
Can we stick with ginkgo since that what most of concourse uses

Reply:
I think using testify for small pluggable stuff is fine, but the core of concourse should be consistent.

Reply:
A long while back we discussed writing new test suites using `testify` instead as it's more accessible for outside contributors and it's generally harder to get wrong. The entire [`worker/backend/` package](https://github.com/concourse/concourse/tree/master/worker/backend) for example uses testify, and in various other places we're trending towards table-driven tests which is really not using a whole lot of Ginkgo anyway.

I don't agree that we should continue using Ginkgo just because the rest of the code uses Ginkgo. If that's the case then we're really just never going to switch. I don't want to end up with the "juicy bits" of Concourse being covered with Ginkgo and the "pluggable bits" being covered with `testify`. And if I had to pick one, after seeing how our test suites tend to grow with Ginkgo, right now I would choose `testify`.

I'm not saying we should go back and re-write all the Ginkgo tests in testify, but as long as we're writing new test suites I don't think we should have to keep using Ginkgo.

Reply:
My understanding was that pluggable bits could/should use testify because there's a greater chance that those will see contributions from the community. 

Maintaining two different styles of test suites for all of concourse is a huge cognitive burden for core contributors and community members. The components of concourse are not well enough isolated to make changes in only one area, so you're going to be making changes across both suites. 

There are definitely cases where we abuse ginkgo, but there's an even greater chance of abusing testify because there is less structure to those tests. All the testitfy tests are new, so they might look better by comparison, but that's just because they haven't had time to suffer the same rot that our ginkgo tests have. You can have a look at dex to see how bad things can get without a good structure in place: https://github.com/concourse/dex/blob/master/server/server_test.go.

Reply:
I kinda get the sense that we're "throwing the baby out with the bath water" because of a couple test suites that got out of hand. I feel like there are lots of other examples of test suites that have aged really well.

Reply:
This isn't about throwing away Ginkgo and switching to Testify because things didn't work out in the past. I don't expect Testify to fix all of our problems.

Here's where I'm at:

* We want more contributors to all parts of Concourse.
* Contributors outside of the Pivotal bubble aren't familiar with Ginkgo.
* Ginkgo is hard to get right, even for the core team.

My point isn't that Testify is better, it's that the choice of Ginkgo is suboptimal for the path we're trying to go. I think Ginkgo is great for sideprojects and walled gardens, but if our primary goal is contributor enablement it hinders more than it helps.

We talked about using Testify for pluggable bits, but that never really happened because we don't work on those bits ourselves. It was a reasonable idea to get the ball rolling, but I don't think that was ever the end goal. Compartmentalizing Testify usage like that sets the tone that we encourage contributions to some components but not others, which feels counterproductive and unrealistic given some of the recent contributions: `set_pipeline` step, `load_var` step, `var_sources:`, etc.

Maintaining Testify suites in addition to our Ginkgo suites in the interim is a burden for *us*, but keep in mind that community members are more likely to be familiar with `testing.T` -style testing. We don't feel a burden right now because we're the ones pushing it on everyone else. The collective burden of us learning Testify is a *lot* lower than the burden of all community contributors - current and future - having to learn Ginkgo. As we add more Testify tests we're at least directly benefiting those in the community, even before we're completely finished with the transition.

Testify takes a day to learn and slightly longer to internalize. It has good matchers, and it has suites for setup and teardown - it just doesn't let you go crazy with it. It provides just enough structure and gets out of the way.

Ginkgo's structure is a double-edged sword. I've seen plenty of test suites become "write-only" because they become harder and harder to refactor as the structure grows and grows. Everything starts off meticulously organized, and over time it gets worse - just like with any code. But by the time it's so bad that people can recognize how bad it is, it's *so much work* to refactor it that no one ever bothers.

To me, the `server_test.go` example you gave just looks like verbose, disorganized code. But it looks pretty easy to refactor.

In any case, we can bring this discussion to the team and check in with how they're feeling about testify on the `containerd` track. It honestly wouldn't take long to change my tests to Ginkgo, since they're table-driven, but to me that feels like a step back. I would rather convert the rest of the package to `testify` and begin adopting it for new code that we add under the `builds/` package.